### ISSUE TITLE
Internal representation question/suggestion

### ISSUE THREAD

**Issue by andyhall** [AdrianGaudebert/entity-system-js#18]
Hi,

I've been further looking into ensy and I have a question/suggestion about its internal representation.

Currently, component data is stored like
```
this.entityComponentData["comp_name"] = {
	"5": { __id:5,  ... },
	"8": { __id:8,  ... }
}
```

Since `entityComponentData` and its values are hashes, I guess a typical System will look like:

```
function updateMeshes() {
	var entHash = entityManager.getComponentsData('has-mesh')
	var ids = Object.keys(entHash)
	for (var i=0; i<ids.length; ++i) {
		var entID = ids[i]
		var data = entList[entID]
		var otherData = entityManager.getComponentDataForEntity('other', entID)
		/* ... */
	}
}
```

Consider, however, if the list of component data was a plain array, accompanied by a hash of IDs:

```
this.entityComponentData["comp_name"] =  {
	ids: {			// hash
		"5": 0,
		"8": 1
	},
	data: [ 		// array
		{ __id:5,  ... },
		{ __id:8,  ... }
	]
}
```
Doing this ought (I haven't tested!) to allow Systems to run a lot faster, at least in v8 (i.e. Chrome and Node). The reason is, Systems only need the data array, and in v8 dealing with plain arrays is much faster than hash tables, provided certain rules are followed (not sparse, starts from 0, etc.). For plain arrays,accessing an element is just a memory read from an offset, but for hash tables it means doing a dictionary lookup of each property name. As a result, looping through an array is much faster than doing `for in` on a hash.

The only user change, I think, is that the user cannot use object keys as entity IDs - they must check the `__id` value instead. But also I think it would be more convenient to get an array, since the System can use `map`, `filter`, etc. E.g.:
```
function updateMeshes() {
	var entArr = entityManager.getComponentsData('has-mesh')
	entArr.map(function(data) {
		var otherData = entityManager.getComponentDataForEntity('other', data.__id)
		/* ... */
	})
}
```
Again, this may be a premature optimization - I have not tried it! But I think that for large projects it would be a noticeable benefit (in Chrome and Node, at least).

Have you considered this, or do you have any thoughts?

Again thanks for the very very nice library!

**Comment by AdrianGaudebert** [comment]
So what you are suggesting is a trade-off. If I were to do that, then indeed it would be faster for Processors to go through the list of components, but then the lookup time when using ``getComponentDataForEntity`` becomes a lot slower, as it has to go through that list to find the correct component to return (instead of just using the entity as the hash key). 

It would be very interesting to run performance tests around this, and around the entire library actually, to be able to verify that such a thing would indeed be an improvement. In theory I do not think the impact is such that it will prevent you from building a game, unless you start doing some crazy things with a huge number of components — in which case I'll be very interested to see what you do and to work with you on improving perfs. :)

In conclusion: I don't think I'll change that until I have evidence that it is an important improvement to make. But I will file an issue to add some performance tests around ``ensy``! 

Thank you for your feedback!

**Comment by andyhall** [comment]
Okay, that's fair enough! If this causes me pain I will write a benchmark and test the savings.

One correction: this wouldn't be a tradeoff, just a straight improvement. The hash of entity ids is to avoid having to search the list. The change to `getComponentDataForEntity` would basically be:
```
--      return this.entityComponentData[componentId][entityId];
++      var cd = this.entityComponentData[componentId];
++      return cd.data[ cd.ids[entityId] ];
```
Or something like that. Basically it's just a little extra bookkeeping that lets you return a faster data structure to the user.

I will play with this more when I have time. If you make a benchmark feel free to tag me, I'd love to see it.

**Comment by andyhall** [comment]
Hey, I went ahead and implemented this change for my own use. It makes the benchmark run about 2-5 times faster, and there are no drawbacks that I know of. (I only tested mac/chrome and FF)

To see the changes check 575980b587cf067bd84a5f61abee88de9b148766, or clone my fork and compare the `data-in-arrays` branch to `master`. 

Note that this is a **breaking change**, which is why I'm not just sending a PR. It's mostly just an internal change, but the external part is that `getComponentsData` now returns an array instead of a hash, and the array indexes are no longer entity IDs. This means that client code processors need to change from doing this:

    var dataObj = manager.getComponentsData('componentName');
    for (var id in dataObj) {
        doSomethingWith( id )
        // ...

to doing this:

    var dataArr = manager.getComponentsData('componentName');
    for (var i=0; i<dataArr.length; ++i) {
        doSomethingWith( dataArr[i].__id )
        // ...

Incidentally it wasn't my goal, but this change also seems to greatly reduce memory thrash, at least in Chrome/v8. 

Old version:
![mem_old](https://cloud.githubusercontent.com/assets/352373/9402808/dc73a560-481b-11e5-859f-06e1be854339.png)

New version:
![mem_new](https://cloud.githubusercontent.com/assets/352373/9402812/e7a2b16a-481b-11e5-86a1-8f1af01556ac.png)

I don't think this is really related to the library, it's probably just due to the processor not needing to call `Object.keys()`, but it's still a nice bonus.

Anyway, if you want to incorporate the change I can clean up and do a PR. Tests are passing, but I have not looked closely at comments, code cleanup or anything like that.

Cheers!
